bundles/org.openhab.binding.monopriceaudio/src/main/java/org/openhab/binding/monopriceaudio/internal/handler/MonopriceAudioHandler.java

```diff
@@ -0,0 +1,754 @@
+            sourcesLabels.add(new StateOption("6", config.inputLabel6));
+
+            // Put the source labels on all active zones
+            List<Integer> activeZones = IntStream.range((1), (numZones + 1)).boxed().collect(Collectors.toList());
+
+            stateDescriptionProvider.setStateOptions(
+                    new ChannelUID(getThing().getUID(), ALL + CHANNEL_DELIMIT + CHANNEL_TYPE_ALLSOURCE), sourcesLabels);
+            activeZones.forEach(zoneNum -> {
+                stateDescriptionProvider.setStateOptions(new ChannelUID(getThing().getUID(),
+                        ZONE + zoneNum.toString() + CHANNEL_DELIMIT + CHANNEL_TYPE_SOURCE), sourcesLabels);
+            });
+
+            // remove the channels for the zones we are not using
+            if (numZones < MAX_ZONES) {
+                List<Channel> channels = new ArrayList<>(this.getThing().getChannels());
+
+                List<Integer> zonesToRemove = IntStream.range((numZones + 1), (MAX_ZONES + 1)).boxed()
+                        .collect(Collectors.toList());
+
+                zonesToRemove.forEach(zone -> {
+                    if (channels.removeIf(c -> (c.getUID().getId().contains("zone" + zone.toString())))) {
+                        logger.debug("Removed channels for zone: {}", zone);
+                    } else {
+                        logger.debug("Could NOT remove channels for zone: {}", zone);
+                    }
+                });
+                updateThing(editThing().withChannels(channels).build());
+            }
+
+            // initialize the all volume state
+            allVolume = initialAllVolume;
+            long allVolumePct = Math
+                    .round((double) (initialAllVolume - MIN_VOLUME) / (double) (MAX_VOLUME - MIN_VOLUME) * 100.0);
+            updateState(ALL + CHANNEL_DELIMIT + CHANNEL_TYPE_ALLVOLUME,
+                    new PercentType(BigDecimal.valueOf(allVolumePct)));
+
+            updateStatus(ThingStatus.UNKNOWN);
+
+            scheduleReconnectJob();
+            schedulePollingJob();
+        }
+
+        logger.debug("Finished initializing!");
+    }
+
+    @Override
+    public void dispose() {
+        logger.debug("Disposing handler for thing {}", getThing().getUID());
+        cancelReconnectJob();
+        cancelPollingJob();
+        closeConnection();
+        super.dispose();
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        String channel = channelUID.getId();
+        String[] channelSplit = channel.split(CHANNEL_DELIMIT);
+        MonopriceAudioZone zone = MonopriceAudioZone.valueOf(channelSplit[0].toUpperCase());
+        String channelType = channelSplit[1];
+
+        if (getThing().getStatus() != ThingStatus.ONLINE) {
+            logger.debug("Thing is not ONLINE; command {} from channel {} is ignored", command, channel);
+            return;
+        }
+
+        if (!connector.isConnected()) {
```

Comment:
Why not just null check connector directly in a few strategic places?

In the code above it would look like this:
if (connector !=null && !connector.isConnected()) {....

Reply:
This code won't be thread-safe. `connector` could be set to null between the null check and invocation of `isConnected()`. The compiler would still complain.

Reply:
In the rotel binding (which this is based off of), they leave the connector variable as non-nullable and then new up the connector with a 'simulated' Impl class when the handler class is instantiated (thus is has a default value and is never null). Then they re-assign it to the desired real Impl (Serial, IP, etc.) during initialize(). Would that acceptable here?

Reply:
I like this approach with simple classes like ScheduledFutures. Be careful when using it with more complex ones when it comes to returning processing results of 'simulated' implementations when actually no processing has have happened.

Reply:
@fwolter... The PR is ready for another look. I cleared up almost all of the warnings. There are a few pesky ones with the config that remain even though config is a local variable that can't be null and the fields are null checked explicitly :(